lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+					else
+						new_range.setMin(rhs_const);
+				}
+				else if (mlir::isa<mlir::rlc::GreaterEqualOp>(statement)){
+					if(result)
+						new_range.setMin(rhs_const);
+					else
+						new_range.setMax(rhs_const-1);
+				}
+				else if (mlir::isa<mlir::rlc::LessEqualOp>(statement)){
+					if(result)
+						new_range.setMax(rhs_const);
+					else
+						new_range.setMin(rhs_const+1);
+				}
+				
+				llvm::outs()<<"should have changed\n";
+				// Insert the range where is needed
+				this->insertOrUpdate(std::make_pair(to_update,branch_to_update),new_range);
+			}
+
+			llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange> ranges;
+		};
+
+		class ConstraintsAnalysis
+				: public mlir::dataflow::DenseBackwardDataFlowAnalysis<
+							ConstraintsLattice>
+		{
+			public:
+			using mlir::dataflow::DenseBackwardDataFlowAnalysis<
+					ConstraintsLattice>::DenseBackwardDataFlowAnalysis;
+
+			// Utility
+			// TODO: understand if i can collapse
+			static bool isArithmetic(mlir::Operation* op){
+				if (mlir::isa<mlir::rlc::AddOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::MultOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::DivOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::SubOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::MinusOp>(op))
+					return true;
+				
+				return false;
+			}
+			static bool isConditional(mlir::Operation* op){
+				if(mlir::isa<mlir::rlc::CondBranch>(op))
+					return true;
+				/*
+				if (mlir::isa<mlir::rlc::GreaterOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::LessOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::GreaterEqualOp>(op))
+					return true;
+				else if (mlir::isa<mlir::rlc::LessEqualOp>(op))
+					return true;
+				*/
+				return false;
+			}
+
+			private:
+
+			// Definitions taken from https://mlir.llvm.org/doxygen/DenseAnalysis_8h_source.html#l00423
+
+			/// Transfer function. Visits an operation with the dense lattice after its
+   			/// execution. This function is expected to set the dense lattice before its
+   			/// execution and trigger propagation in case of change..
+			void visitOperation(
+				mlir::Operation *op, 
+				const ConstraintsLattice &after,
+                ConstraintsLattice *before) override
+			{
+				llvm::outs()<<"visiting operation: ";
+				op->print(llvm::outs());
+				llvm::outs()<<"\n";
+				// For the moment implement the basic operations
+				auto res=before->join(after);
+				
+				// We have to decide which return values to update (true or false)
+				// In order to do so we need to :
+				// - peek at the successors of this block 
+				// - look at their lattice
+				// - find in their lattice if there is some value with key true/false
+				// - keep in mind that information
+
+				if (mlir::rlc::ConstraintsAnalysis::isArithmetic(op)){
+					bool can_true=false;
+					bool can_false=false;
+
+					// Since here we have a lattice that is copied we can just look at it
```

Comment:
you are doing this looping check a lot, can you promote it to a function?

Reply:
Sure